### PR TITLE
Feature: Vouchers and promotions using Voucherify

### PR DESCRIPTION
**Pull request by weronika-kurczyna**
## Background

As we established in a recent meeting between Voucherify and Orium, we were asked to create two separate pull requests. 
This is the second one, containing an extension of the current behavior of the app with the functionality of discounts and promotions operated by **Voucherify**.
This PR contains changes that, after running the appropriate script, will be applied to the application, which will enable the use of Voucherify in ComposableUI.

## Setting up Voucherify integration

All files necessary to properly configure the Voucherify integration are located in `scripts/voucherify-setup`.

To find setup details, go to `docs/docs/integrations/promotions/voucherify.md` (**Integrating Voucherify with Composable UI** section)

## What was done? (code changes)

The changes we made are located in:
- `packages/voucherify`,
- `scripts/voucherify-setup`,
- `docs/docs/integrations/promotions`.

1. Documentation regarding the integration is located in: `docs/docs/integrations/promotions`
2. Configuration is located in: `scripts/voucherify-setup`
3. Implementation of Voucherify is located in: `packages/voucherify`. Here are functions that extend the default behavior and allow the use of Voucherify features.

## Using Voucherify features
After activating Voucherify (`pnpm voucherify-acivate`), you should be able to create some exemplary campaigns, such as:
- Discount coupons: amount / percent discount for whole cart or specific product/products
  - Bulk codes (specific number of codes generated based on a pattern)
  - Standalone code (with custom name)
- Promotions
  - amount discount
  - percent discount
  - fixed amount
  Each of the above may have defined [Validation Rules](https://support.voucherify.io/article/148-how-to-build-a-rule), to apply the discount only under certain conditions.
  
Below is a promotional campaign with three different promotion tiers:
- 5$ discount (apply to whole cart)
- 5% discount (apply to whole cart) - when quantity of items is greater than 2
- 20$ discount (when total order amount is greater than 300$)
image
In addition, a voucher `ONE_PRODUCT` has been created that provides a 10% discount on the `Essential Pajama set in slate` product

A video was recorded to demonstrate how the above discounts work:

https://github.com/composable-com/composable-ui/assets/117282008/92d6399e-23e0-4ee6-9666-f0c1c8f5228b

### CODE REVIEW

**Comment by dannytlake** [comment]
@marcin-slezak @weronika-kurczyna please rebase `voucherifyio:feat/voucherify-integration` on the latest from `composable-ui:main` to clean up this PR (this should remove all the changes under `/compoable-ui` from the PR)

File: docs/docs/integrations/promotions/voucherify.md [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+To configure product base promotions in Voucherify, propagate product definitions to your Voucherify account:
+
+1. Open the terminal and navigate to the `scripts` directory.
+2. In the `scripts` directory, run the following command:
+  ```
+  pnpm install
+  ```
+3. To set up Voucherify, run the following command:
+  ```
+  pnpm voucherify-setup
```
Comment (dannytlake) [inline docs/docs/integrations/promotions/voucherify.md:92]: @marcin-slezak @weronika-kurczyna I think should this be `pnpm voucherify-preconfigure`? There is no script `voucherify-setup`

File: docs/docs/integrations/promotions/voucherify.md [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+1. [Create a Voucherify account](https://app.voucherify.io/#/signup).
+2. In Voucherify Dashboard, [set Discount Application Rule to "Partial"](https://support.voucherify.io/article/604-stacking-rules#application-rules)
+   Please go to:
+- `Redemptions` tab on the left side of the Voucherify dashboard
+- Select `Stacking rules`
+- In `Discounts' application rules` select `Edit settings`
+- Change the `Application rule` to `Partial - inapplicable (not met criteria) discounts are omitted`
+3. Retrieve your API keys from your Voucherify dashboard and set the following environment variables:
+
+```code
```
Comment (dannytlake) [inline docs/docs/integrations/promotions/voucherify.md:72]: Please add a note about adding these env vars to both the `composable-ui` app as well as in `scripts/.env`

File: scripts/package.json [hunk 3]
```diff
@@ -3,12 +3,16 @@
   "version": "0.0.0",
   "private": true,
   "scripts": {
-    "algolia-setup": "ts-node src/algolia-setup/index.ts"
+    "algolia-setup": "ts-node src/algolia-setup/index.ts",
+    "voucherify-preconfigure": "ts-node src/voucherify-setup/index.ts",
```
Comment (dannytlake) [inline scripts/package.json:7]: @marcin-slezak @weronika-kurczyna I get an error when running this against my Voucherify sandbox:

image

File: docs/docs/integrations/promotions/voucherify.md (hunk 1 above)
Comment (weronika-kurczyna) [reply docs/docs/integrations/promotions/voucherify.md]: You are right, the wrong name remained in the documentation. Fixed.

File: scripts/package.json (hunk 3 above)
Comment (weronika-kurczyna) [reply scripts/package.json:7]: It looks like your trial has expired. i did not put information in the documentation regarding that adding products is optional, it is not required for proper validation and redemption. So you can use the application without enabling this script - i have updated the documentation. if you need an extended trial on this account let us know, we will probably be able to activate it.

**Comment by weronika-kurczyna** [comment]
@dannytlake This PR is ready for review

File: docs/docs/integrations/promotions/voucherify.md [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+Running a script that adds products to Voucherify is optional. Without running it, you will still be able to perform validation and redemption.
+However, it may be required in the future if other types of discounts are implemented.
+
+**Note:** If you are using the free plan and sandbox project, keep in mind that it has a limit of 100 API calls per hour. 
+With the basic plan you should be able to perform the `voucherify-preconfigure` action without any problems. 
+However, on subsequent attempts to add products, you may get an error: `429 - TOO_MANY_REQUESTS`.
+In this situation, you will only be able to make further attempts in the next hour. You can also use the app without adding products.
+
+
+> **_NOTE:_**  If you are using the Free plan and Sandbox project, keep in mind that it has a limit of 100 API calls per hour.
```
Comment (dannytlake) [inline docs/docs/integrations/promotions/voucherify.md:106]: @weronika-kurczyna this NOTE looks to be a duplicate of the Note that starts on line 100. May want to remove one of these.

image

**Review by dannytlake — APPROVED** [review]
Please see comment regarding duplicate Note in `docs/docs/integrations/promotions/voucherify.md`

File: docs/docs/integrations/promotions/voucherify.md (hunk 4 above)
Comment (weronika-kurczyna) [reply docs/docs/integrations/promotions/voucherify.md:106]: Ah, sorry for the confusion. It has been corrected.

Comment (dannytlake) [reply docs/docs/integrations/promotions/voucherify.md:100]: Thank you!

**Pull request merged**